readme
Lanner Bypass/Watchdog Module Access Code
=========================================

Aug 31, 2018

Contents
========
- In This Release
- Building 
- Installation
- HOW-TO Access
- Support model
- License


In This Release
===============
This file describes the Lanner bypass/watchdog module acces code.  
This package provides program and driver for both DOS and Linux.
Separated makefile available as well.(Makefile.dos or Makefile.linux)

For DOS environment, Use DJGPP as compiler, use DJGPP and Makefile.dos.

For Linux, supports kernel versions 2.4.x/2.6/3.x/4.x and use Makefile.linux. 


This release provide 3 ways to access Lanner bypass/watchdog module: 

1. DIRECT_IO_ACCESS:

   To access Lanner bypass/watchdog module by userland application direct
   accessing. This mode JUST FOR providing quick way for user to realize 
   Lannner watchdog/bypass function. This mode supported on DOS and Linux.

   Note: This mode is enabled by set DIRECT_IO_ACCESS=1 in Makefile.


2. LANNER_DRIVER:

   To access Lanner bypass/watchdog module through lanner driver. 
   Lanner driver(bpwd_drv.ko) is available in bin sub-directory after make.
   This driver is only supported as a loadable module at this time.  Lanner is
   not supplying patches against the kernel source to allow for static linking
   of the driver. 

   Note: This mode is enabled by set DIRECT_IO_ACCESS=0 and LANNER_DRIVER=1 in 
   Makefile.


3. LINUX_KERNEL_DRIVER:

   To access Lanner bypass/watchdog module through linux i2c dev interface. 
   i2c-dev.ko and i2c-i801.ko are need to be loader for normal access.
   /dev/i2c-0 c 89 0 node is need to be created as well.

   Note1: This mode is enabled by set DIRECT_IO_ACCESS=0 and LANNER_DRIVER=0 in 
          Makefile.
   Note2: Linux kernel driver(i2c-i801.c) is lacked for block accessing. 
          Varied kernel version may get different results. 
          For example, 2.6.18(CentOS-54) do not support SMBUS block function.
	  2.6.25(FC9) block write function fail. 
          Since Lanner bypass/watchdog ARP function relied on BLOCK+PEC 
          accessing, this cause ARP function fail.
   Note3: MB-5330 is using AMD Structure system, and we only support 
          DIRECT_IO_ACCESS and LANNER_DRIVER in source code.
          


Building
========
To build, 4 steps to completed:

1. Identify current OS and select proper Makefile. 
   Copy Makefile.(os) to Makefile.

2. Select access mode.(DIRECT_IO_ACCESS=[0|1] and LANNER_DRIVER=[0|1] by 
   edit Makefile)

4. Select Major number assign mode. (MAJOR_NUMBER=[0|xxx] by edit Makefile)

	 Note: 1. MAJOR_NUMBER define only usefule in LANNER_DRIVER=1
         2. Set MAJOR_NUMBER=0, then use Dynamic major number assign
	       

3. Just type make to build, once completed, bin sub-directory contains.



Installation
============
To installation, depend on what's access mode you set:

If DIRECT_IO_ACCESS=1, no driver is need. Just execute bpwd_tst program to
to access Lanner bypass/watchdog module.

If DIRECT_IO_ACCESS=0 and LANNER_DRIVER=1, driver is needed. Insert module and
create node in /dev as below example:

	#insmod bpwd_drv.[k]o

	#mknod /dev/bpwd_drv c 247 0

  Note: '247' in here is for example, if '247' is already occupied,
           please check other major number of kernel driver populated. 
           or modify Makefile's "MAJOR_NUMBER" to "0" use Dynamic number,
           then chenag it as dynamic assigned unmber.

If DIRECT_IO_ACCESS=0 and LANNER_DRIVER=0, linux kernel driver is needed. 
i2c subsystem: i2c-core, i2c-dev and busses/i2c-i801 need to be loaded before
access Lanner bypass/watchdog module, dev node need to create as below:

	#mknod /dev/i2c-0 c 89 0
	 
	Note: '0' in here is for example, check proper bus number of kernel driver 
              populated. 
              

HOW-TO ACCESS
=============
Once build completed, application(and driver) is available in bin sub-directory.
Main access program is bpwd_tst. just run it without argument will list help.

Lanner Bypass/Watchdog Access Code:V1.2.5 2018-04-19
Usage: ./bpwd_tst -S [-M [model_name]]
	Or 
	./bpwd_tst -S [-M [Simple_model_name]]
        Scan Lanner bypass module.

Usage: ./bpwd_tst -r -d [i2c_address] -c [command] [-M [module_name]]
	Or
	./bpwd_tst -r -d [i2c_address] -c [command] [-M [Simple_model_name]]
        SMBUS read word data protocol

Usage: ./bpwd_tst -w -d [i2c_address] -c [command] -o [byte_data] [-M [module_name]]
	Or
	./bpwd_tst -w -d [i2c_address] -c [command] -o [byte_data] [-M [Simple_model_name]]
        SMBUS write byte data protocol

Usage: ./bpwd_tst -I -d [i2c_address] [-M [module_name]]
	Or
	./bpwd_tst -I -d [i2c_address] [-M [Simple_model_name]]
        Get Lanner bypass module information
Global argument:
        -v: be verbose
        -h: help


For multiple card configuration
-------------------------------
  Since multiple bypass add-on card maybe populate on motherboard, isolate bypass 
card by I2C switch is implemented in following motherboard:

	model_name	simple_model_name	Lanner model 
	=====================================================
	MB-887X		Mode1			  MB-8875, MB-8876, MB-8865, MB-8866, 
	                      MB-8877, MB-8771,NCB-5210
	MB-8877EXT	Mode2			MB-8877(with RC-8877 expansion card, 4 slots)							
	MB-8865EXT	Mode3			MB-8865(with RC-8865 expansion card, 6 slots)						
	MB-7582EXT	Mode4			MB-7582(with RC-7582 expansion card, 6 slots)								
	MB-7583		  Mode5			MB-7583, MB-7573								
	MB-8771EXT	Mode6			MB-8771(with RC-8771 expansion card, 2 slots)
  MB-8895		  Mode7			MB-8896, MB-8895, MB-8893, NCB-5510A
	MB-MFI20K	  Mode8			MB-MFI20K(with RC-MFI20K expansion card, 9 slots)  	
  MB-9655		  Mode9			MB-9655
	NCB-5510EXT	Mode10		NCB-5510(with RC-88771 expansion card, 8 slots)
	NCB-5220	  Mode11		NCB-5220
	NCB-2510	  Mode12		NCB-2510
	NCB-6210	  Mode13		NCB-6210
 
Program need to specific with -M [model_name] or -M [simple_model_name]
for dedicate motherboard and device address [specific by -d]
 is fixed as below:
	0x30: slot1
	0x31: slot2
	0x32: slot3
	0x33: slot4
	0x34: slot5
	0x35: slot6
	0x36: slot7
  0x37: slot8
  0x38: slot9

If bypass controller is add-on card, below using i2c address 0x30 at first slot on the motherboard. 
Following is example within -M command:

  MB-8875 as example, access slot1( address is :0x30)
    Write:
	bpwd_tst -w -d 0x30 -c 0x12 -o 0x03 -M MB-887X 
		Or
	bpwd_tst -w -d 0x30 -c 0x12 -o 0x03 -M Mode1
	
    Read:
	bpwd_tst -r -d 0x30 -c 0x12 -M MB-887X  
		Or
	bpwd_tst -r -d 0x30 -c 0x12 -M Mode1
	
     Scan Bus:
	bpwd_tst -S -M MB-887X
		Or
	bpwd_tst -S -M Mode1
	
For on-board configuration
--------------------------
If bypass controller is embedded in motherboard, no -M parameter is needed and i2c address is fixed to 0x37.
Following is example without -M command:

     MB-8765 as exampple
     Write
	bpwd_tst -w -d 0x37 -c 0x12 -o 0x03
     Read:
	bpwd_tst -r -d 0x37 -c 0x12
     Scan Bus:
	bpwd_tst -S

Example sub-directory presented you how to access module. 
You may add -M parameter for proper motherboard.



Support model
=============
Following is the list of Lanner's model which support 3rd bypass.

3rd bypass add-on card, which connect to motherboard with i2c switch to control each device.
You should have to use -M command to control each slot.
	model_name	simple_model_name	Lanner model 
	=====================================================
	MB-887X		  Mode1			MB-8875, MB-8876, MB-8865, MB-8866,
	                      MB-8877, MB-8771,NCB-5210
	MB-8877EXT	Mode2			MB-8877(with RC-8877 expansion card, 4 slots)							
  MB-8865EXT	Mode3			MB-8865(with RC-8865 expansion card, 6 slots)								
	MB-7582EXT	Mode4			MB-7582(with RC-7582 expansion card, 6 slots)								
	MB-7583		  Mode5			MB-7583, MB-7573								
	MB-8771EXT	Mode6			MB-8771(with RC-8771 expansion card, 2 slots)
  MB-8895		  Mode7			MB-8896, MB-8895, MB-8893, NCB-5510A
	MB-MFI20K	  Mode8			MB-MFI20K(with RC-MFI20K expansion card, 9 slots)  	
  MB-9655		  Mode9			MB-9655
	NCB-5510EXT	Mode10  	NCB-5510(with RC-88771 expansion card, 8 slots)	
	NCB-5220	  Mode11		NCB-5220
	NCB-2510	  Mode12		NCB-2510
	NCB-6210	  Mode13		NCB-6210
	
3rd bypass embedded on motherboard, which only use i2c address 0x37 to control the micro-processor.
No need to use -M command to control micro-processor.       	 
    Lanner model
    ====================================
    MB-8756
    MB-8770
    MB-7582
    MB-5330	( There is only support lanner driver and direct i/o access to MB-5330 )	
    MB-7583
    FW-8771

License
=======
Lanner bypass/watchdog module access code
Copyright(c) 2013 Lanner Electronics Inc.

This program is free software; you can redistribute it and/or modify it
under the terms and conditions of the GNU General Public License,
version 2, as published by the Free Software Foundation.

This program is distributed in the hope it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.

The full GNU General Public License is included in this distribution in
the file called "COPYING".







history.txt
Version history

Version	Date            Description
======= ========================================================================
v1.2.9	May 03,2019	1. Add new model support for NCB-6210
			2. Fix kernel driver mode i2c guess bus wrong issue
			3. Set default mode is kernel driver (Makefile.linux)

v1.2.8  Aug 31,2018     1. Fix v1.2.7 Lanner driver mode probe SMbus device error.
			   if PCI device not exist than do read pci reg will cause system error. 
			   So modify source code to check pci device exist then do the class code check.
			   
v1.2.7  Aug 16,2018	1. Add Mode11(NCB-5220) and mode12(NCB-2510) Support
			2. Fix Lanner driver mode probe SMbus device not correct issue(NCB-5220)
			3. Enlarge SMBUS timeout setting
			
v1.2.6  Jul 24.2018     1. Add Dynamic Major number support
                        2. fix kernel 4.13/4.17 compile error issue.
			3. fix compile warring message
			4. fix can't access SMbus in ubuntu OS, if DIRECT_IO_ACCESS is set 1.
			 
v1.2.5  Apr 19.2018	1. Add NCB-5510EXT model_name
			2. Add support NCB-5210(use -M MB-887X).
			3. Add 0x15 command(Get/Set the time unit of watchdog timer)
			   P.S The 0x15 command need firmware code support.
			4. Add simple_model(Mode1 to Mode10)
			5. Modify 0x7/0x8/0x9/0x22/0x32/0x42 command output text, 
			   delete 'second' text,only show value.
			6. Modify readme, add the below:
			 [ Note: '247' in here is for example, if '247' is already occupied,
            			check other major number of kernel driver populated.]
			 and Description item 1/2/4 content

v1.2.4  Mar 14.2017	1.  1. Modify /driver/Makefile to suit for linux kernel
                             version over 4.x.x

v1.2.3  Jan 2.2016	1. Add support apollo lake platform 

v1.2.2  Sep 1.2015	1. Add support skylake platform      

v1.2.1  Feb 21.2014	1. Add MB-8771EXT model_name for 
                                        MB-8771EXT with RC-8771EXT expansion card, 2 slots.

v1.2.0  Oct 19.2013	1. Add MB-8877EXT model_name for 
                                         MB-8877EXT with RC-8877EXT expansion card, 4 slots.

v1.1.9  Jul 18.2013	1. Add MB-MFI20K model_name for 
                                        MB-MFI20K with RC-MFI20K expansion card, 9 slots.
			2. Add MB-8877 model_name , support MB-8771 2 slots.

v1.1.8  Jan 10,2013	1. Add MB-7583 model_name 

v1.1.7  Dec 20,2012	1. Fix unresolved symbol pci_enable_device_io in kernel 
                                      -2.4.x 
 
v1.1.6  Dec 07,2012	1. Add MB-7582EXT model_name for 
                                       MB-7582 with RC-75822Z+RC-75824Z expansion card,
                                       6 slots.
                                  2. Printout support model_name in help.
 
v1.1.5  Aug 14,2012	1. Fix "-I" and "-S" command failed work with "-M".
                                      2. Just scan lanner bypass modules in -S command.
                                      3. Add MB-8865EXT model_name for 
                                         MB8865 with RC-8865 expansion card, 6 slots.
                                     4. Printout detail error message while SMBus Error

v1.1.4  May 29,2012     1. Add new support for MB-8866 within 4 pcie slot, and 
                                      using model name MB-887X.

v1.1.3  May 21,2012     1. Modify /driver/Makefile to suit for linux kernel 
                                         version over 3.x.x
                                      2. Add support for AMD sturcture MBB-5330
                                      3. lanner driver : claim resource -> to prevent kernel
                                          and lanner driver co-exis bug.
                                      4. linux kernel driver : add i2c bus adapter.

v1.1.2  Mar 28,2012     1. Change LCM driver major number to 247

v1.1.1  Mar 22,2012     1. Change LCM driver major number to 251

v1.0.9  Jan 17,2012     1. Add support for MB-8895 with maximum 8 PCIE slots. 

v1.0.8  Sep 30,2011     1. Fix the error of NULL value as a parameter to pointer
                                        in i2c_smbus_write_byte function.
 
v1.0.7  Jan 05,2011     1. Fix i2c bus scan function [ bpwd_tst -S ] return 
                                        value error

v1.0.6  Dec 16,2010     1. Define CPLD onboard i2c address to 0x37
                                         Define CPLD on lan module(ex:IG-425) i2c address to 
                                         0x30

v1.0.5  Dec 9, 2010     1. Fix compiler error in Linux 2.6
                                    2. correct version information

v1.0.4  Nov 30, 2010    1. Support linux kernel 2.4 
                                          [LANNER DRIVER]: work
	                          [LINUX DRIVER]: not test yet 
                                          [DIRECT IO ACCESS]: work

v1.0.3  Nov 25, 2010    1. Adjust parameter -M usage :
                                         If there are I2C Switch on the board, user need to 
                                         add -M command with parameter [module_name] to 
                                         switch i2c channel.
	                        Current support [module_name]: 
                                        MB-887X(MB-8875 & MB-8876) , MB-9655
			    ex:
			     bpwd_tst -d 0x30 -c 0x12 -o 0x03 -M MB-887X -w
							
V1.0.2	Oct. 1, 2010    1. For various motherboard design, I2C bus may be 
                            isolated from the world. Adding new parameter -M to 
                            specific model name's method for going through.
                            Current support model: MB-8876, see README for
                            command usage.

V1.0.1	Sep 16, 2010    1. Fix compiler error in DJGPP
                        2. Fix typo WDBP_MAJOR in driver/bpwd_main.c
                        3. Remove ARP support	

include/ioaccess.h
/*******************************************************************************

  ioaccess.h : Low level register definition file for Lanner bypass/watchdog 
		module access code

  Lanner bypass/watchdog module access code
  Copyright(c) 2010 Lanner Electronics Inc.

  This program is free software; you can redistribute it and/or modify it
  under the terms and conditions of the GNU General Public License,
  version 2, as published by the Free Software Foundation.

  This program is distributed in the hope it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
  more details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc.,
  51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.

  The full GNU General Public License is included in this distribution in
  the file called "COPYING".

*******************************************************************************/


#include "bp_def.h"

#ifndef DJGPP
#define inportb(x) inb(x)
#define outportb(x, y) outb(y, x)
#define inportl(x) inl(x)
#define outportl(x, y) outl(y, x)
#endif

/* SMBUS timeout setting */
//-v1.2.7 #define MAX_TIMEOUT             1000
#define MAX_TIMEOUT             5000

/* Intel 801 SMBus address offsets */
#define SMBHSTSTS       (0)
#define SMBHSTCNT       (2)
#define SMBHSTCMD       (3)
#define SMBHSTADD       (4)
#define SMBHSTDAT0      (5)
#define SMBHSTDAT1      (6)
#define SMBBLKDAT       (7)
#define SMBPEC          (8)         /* ICH3 and later */
#define SMBAUXSTS       (12)        /* ICH4 and later */
#define SMBAUXCTL       (13)        /* ICH4 and later */

/* I801 Hosts Status register bits */
#define SMBHSTSTS_BYTE_DONE     0x80
#define SMBHSTSTS_INUSE_STS     0x40
#define SMBHSTSTS_SMBALERT_STS  0x20
#define SMBHSTSTS_FAILED        0x10
#define SMBHSTSTS_BUS_ERR       0x08
#define SMBHSTSTS_DEV_ERR       0x04
#define SMBHSTSTS_INTR          0x02
#define SMBHSTSTS_HOST_BUSY     0x01

/* kill bit for SMBHSTCNT */
#define SMBHSTCNT_KILL          2

/* SMBUS ARP assigned base address */
#define ARP_BASE_ADDR   0x20

/* SMBUS 2.0 ARP UUID definition */
#define ARP_UUID_VID0   2   //Vendor ID
#define ARP_UUID_VID1   3   //Vendor ID
#define ARP_UUID_VSID0  12  //Vendor Specific ID


typedef struct {
	unsigned char bp_addr;
	unsigned char bp_cmd;
	unsigned char bp_pec;
	unsigned char bp_inhibit_err;
	unsigned char bp_buffer[20];
} bp_s;

#define I2C_READ        0x01
#define I2C_WRITE       0x00
/* return 0 mean host ready for transfer, otherwise return -1 */
int host_not_ready(int fd);

/* wait command processed and check error occured */
int host_busy_or_err(int fd, int inhibit_err);
int i2c_smbus_write_byte(int fd, unsigned char i2c_address,
                        unsigned char i2c_command,
                        int pec, unsigned char *pec_byte);
int i2c_smbus_write_byte_data(int fd, unsigned char i2c_address,
                                unsigned char i2c_register, unsigned char value);

int i2c_smbus_read_byte_data(int fd, unsigned char i2c_address,
                             unsigned char i2c_register, unsigned char *ret_val, int inhibit_err);

int i2c_smbus_read_word_data(int fd, unsigned char i2c_address,
                            unsigned char i2c_register, unsigned short *ret_val);

int i2c_smbus_read_block_data(int fd, unsigned char i2c_address,
                            unsigned char i2c_register, unsigned char ret_array[],
                            unsigned char *pec_byte);

int i2c_smbus_write_block_data(int fd, unsigned char i2c_address,
                            unsigned char i2c_register, unsigned char ret_array[],
                            unsigned char *pec_byte);




include/bp_def.h
/*******************************************************************************

  bp_def.h : command definition file for Lanner bypass/watchdog module access code

  Lanner bypass/watchdog module access code
  Copyright(c) 2010 Lanner Electronics Inc.

  This program is free software; you can redistribute it and/or modify it
  under the terms and conditions of the GNU General Public License,
  version 2, as published by the Free Software Foundation.

  This program is distributed in the hope it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
  more details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc.,
  51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.

  The full GNU General Public License is included in this distribution in
  the file called "COPYING".

*******************************************************************************/

/* Return code definitions */
#define CMD_SUCCESS 0x80

/* command definitions */

/* CMD = 0x01: major version */
#define MAJOR_VER_CMD       0x01

/* CMD = 0x02: minor version */
#define MINOR_VER_CMD       0x02

/* CMD = 0x03: module capability */
#define MODULE_CAP_CMD      0x03
/*      bit definition */
#define MODULE_CAP_SO_BIT   0x01
#define MODULE_CAP_JO_BIT   0x02
#define MODULE_CAP_RT_BIT   0x04
#define MODULE_CAP_WDT1_BIT 0x08
#define MODULE_CAP_WDT2_BIT 0x10
#define MODULE_CAP_WDT3_BIT 0x20

/* CMD = 0x04: System-Off bypass capability */
#define SO_BP_CAP_CMD       0x04
/*      variable definition */
#define BYPASS_CAP_1PAIR    0x01
#define BYPASS_CAP_2PAIR    0x03
#define BYPASS_CAP_3PAIR    0x07
#define BYPASS_CAP_4PAIR    0x0F
#define BYPASS_CAP_5PAIR    0x1F
#define BYPASS_CAP_6PAIR    0x3F
#define BYPASS_CAP_7PAIR    0x7F
#define BYPASS_CAP_8PAIR    0xFF

/* CMD = 0x05: Just-On bypass capability */
#define JO_BP_CAP_CMD       0x05

/* CMD = 0x06: Run-Time bypass capability */
#define RT_BP_CAP_CMD       0x06

/* CMD = 0x07: watchdog 1(Run-Time) capability */
#define RT_WDT1_CAP_CMD     0x07

/* CMD = 0x08: watchdog 2(Run-Time) capability */
#define RT_WDT2_CAP_CMD     0x08

/* CMD = 0x09: watchdog 3(Just-On) capability */
#define JO_WDT3_CAP_CMD     0x09
    
/* CMD = 0x0A: Load factory default */
#define LOAD_DEFAULT_CMD    0x0A
    
/* CMD = 0x0B: Save current setting */
#define SAVE_SETTING_CMD    0x0B
    
/* CMD = 0x0C: Board ID */
#define BOARD_ID_CMD        0x0C

/* CMD = Get or Set the time unit of watchdog timer */
#define WDT_TIME_UNIT_CMD   0x15       //Tim 20180123

/*================ Bypass ==================*/
    
/* CMD = 0x10: System-Off bypass setting */
#define SO_BP_SET_CMD       0x10
/*      bit definition */
#define BYPASS_PAIR_1_BIT 0x01
#define BYPASS_PAIR_2_BIT 0x02
#define BYPASS_PAIR_3_BIT 0x04
#define BYPASS_PAIR_4_BIT 0x08
#define BYPASS_PAIR_5_BIT 0x10
#define BYPASS_PAIR_6_BIT 0x20
#define BYPASS_PAIR_7_BIT 0x40
#define BYPASS_PAIR_8_BIT 0x80

/* CMD = 0x11: Just-On bypass setting */
#define JO_BP_SET_CMD       0x11

/* CMD = 0x12: Run-Time bypass setting */
#define RT_BP_SET_CMD       0x12

/*================ Watchdog 1 (Run-Time) ==================*/

/* CMD = 0x20: watchdog1(Run-Time) status */
#define RT_WDT1_STS_CMD     0x20
/*      variable definition */
#define WDT_STS_STOP        0x00
#define WDT_STS_RUNNING     0x01
#define WDT_STS_EXPIRED     0x02

/* CMD = 0x21: watchdog1(Run-Time) pair setting for timeout */
#define RT_WDT1_PAIR_CMD    0x21

/* CMD = 0x22: watchdog1(Run-Time) timer counter */
#define RT_WDT1_TIMER_CMD   0x22

/* CMD = 0x23: watchdog1(Run-Time) estimated time */
#define RT_WDT1_ESTIMATE_CMD 0x23

/* CMD = 0x24: watchdog1(Run-Time) start */
#define RT_WDT1_START_CMD    0x24

/* CMD = 0x25: watchdog1(Run-Time) stop */
#define RT_WDT1_STOP_CMD    0x25

/*================ Watchdog 2 (Run-Time) ==================*/

/* CMD = 0x30: watchdog2(Run-Time) status */
#define RT_WDT2_STS_CMD     0x30

/* CMD = 0x31: watchdog2(Run-Time) pair setting for timeout */
#define RT_WDT2_PAIR_CMD    0x31

/* CMD = 0x32: watchdog1(Run-Time) timer counter */
#define RT_WDT2_TIMER_CMD   0x32

/* CMD = 0x33: watchdog2(Run-Time) estimated time */
#define RT_WDT2_ESTIMATE_CMD 0x33

/* CMD = 0x34: watchdog2(Run-Time) start */
#define RT_WDT2_START_CMD    0x34

/* CMD = 0x35: watchdog2(Run-Time) stop */
#define RT_WDT2_STOP_CMD    0x35

/*================ Watchdog 3 (Just-On) ==================*/

/* CMD = 0x40: watchdog3(Just-On) status */
#define JO_WDT3_STS_CMD     0x40

/* CMD = 0x41: watchdog3(Just-On) pair setting for timeout */
#define JO_WDT3_PAIR_CMD    0x41

/* CMD = 0x42: watchdog3(Just-On) timer counter */
#define JO_WDT3_TIMER_CMD   0x42

/* CMD = 0x43: watchdog3(Just-On) estimated time */
#define JO_WDT3_ESTIMATE_CMD 0x43

/* CMD = 0x45: watchdog3(Just-On) stop */
#define JO_WDT3_STOP_CMD    0x45



